docs/sha2-256-chunked.md

```diff
@@ -0,0 +1,43 @@
+# sha2-256-chunked
+
+## DRAFT multihash codec 0xb510
+
+> Hash of concatenated SHA2-256 digests of 8*2^n MiB source chunks
+> where n = log(source_size / (10^4 * 8MiB))
+
+This variant of sha2-256 is designed to enable very large files
+to be efficiently uploaded and hashed in parallel using fixed size chunks
+(8MiB to start with), with the final result being a "top hash"
+that doesn't depend on the upload order.
+
+The algorithm has an upper limit of 10,000 hashes.
+If the file is larger than 80,000 MiB,
+it will double the block size until the number of blocks
+is under that limit.
+
+```pseudocode
+n = floor(log(source_size / (10_000 * 8 MiB)))
+chunk_size = 8*2^n MiB
+```
```

Comment:
that doesn't actually work
I suggest to use code that we actually use and that is tested
if for some reason we don't want to use the existing iterative algorithm these are my best guesses:

```
MIN_CHUNK_SIZE = 8 * 2 ** 20
MAX_NUM_CHUNKS = 10_000

from math import ceil, log2

def get_chunk_size1(source_size: int):
    if source_size == 0:
        return MIN_CHUNK_SIZE
    return MIN_CHUNK_SIZE * 2 ** ceil(log2(ceil(source_size / (MAX_NUM_CHUNKS * MIN_CHUNK_SIZE))))
def get_chunk_size2(source_size: int):
    if source_size == 0:
        return MIN_CHUNK_SIZE
    return MIN_CHUNK_SIZE * 2 ** ((source_size - 1) // (MAX_NUM_CHUNKS * MIN_CHUNK_SIZE)).bit_length()
```

Reply:
Just commented above, but I think ceil(log2(...)) is best.

Reply:
Updated with Dima's comment.
```
ceil(log2(ceil(source_size...)))
``
@sir-sigurd Why do we need the double `ceil`? Is that actually necessary for correctness, or just to match the way we do doubling?

Reply:
@drernie 

seems wrong, doesn't it?

```python
In [55]: MIN_CHUNK_SIZE = 8 * 2 ** 20
    ...: MAX_NUM_CHUNKS = 10_000
    ...:
    ...: from math import ceil, log2
    ...:
    ...: def get_chunk_size(source_size: int):
    ...:     return MIN_CHUNK_SIZE * 2 ** ceil(log2(source_size / (MAX_NUM_CHUNKS * MIN_CHUNK_SIZE)))
    ...:

In [56]: get_chunk_size(0)
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
Cell In[56], line 1
----> 1 get_chunk_size(0)

Cell In[55], line 7, in get_chunk_size(source_size)
      6 def get_chunk_size(source_size: int):
----> 7     return MIN_CHUNK_SIZE * 2 ** ceil(log2(source_size / (MAX_NUM_CHUNKS * MIN_CHUNK_SIZE)))

ValueError: math domain error

In [57]: get_chunk_size(1)
Out[57]: 0.0001220703125
```

Reply:
Technically, we need a max(0, log2(...)) to avoid <8MB chunks. And something similar to avoid log2(0). @drernie - is it worth the complexity here? We'd need something like:

```
n = floor(max(0, log(max(1, source_size / (10_000 * 8 MiB)))))
```

Reply:
Anyways, we can fix it later.